### PR TITLE
fixes #21505 (overload resolution of explicit constructors for imported C++ types)

### LINKED ISSUE #21505: importcpp types with explicit constructor "producing chosen constructor is explicit in copy-initialization"
### Description

Codegen is now producing an extra `{}` for cpp types which breaks types with explicit constructors 
The change is originated here: https://github.com/nim-lang/Nim/pull/21378/files

Minimal repro

```nim

{.emit:"""/*TYPESECTION*/
struct FVector4
{
    int a;
    explicit FVector4(int inA = 0) {}  
};

""".}

type FVector4 {.importcpp.} = object
    a : int32



proc vector4_Zero*(): FVector4 {.exportcpp: "$1_" .} =
  type
    Params  = object
      returnValue : FVector4 = FVector4()
  var param   = Params()
  return param.returnValue
```

### Nim Version

1.6.9

### Current Output

```text
/Volumes/Store/Dropbox/Code/Projects/nimlab/lab/nimcache/@mcppinitissue.nim.cpp:81:48: error: chosen constructor is explicit in copy-initialization
        tyObject_Params__JEe9afTCYqDowNJPGU6j9a3Q T1_{};
                                                      ^
/Volumes/Store/Dropbox/Code/Projects/nimlab/lab/nimcache/@mcppinitissue.nim.cpp:35:14: note: explicit constructor declared here
    explicit FVector4(int inA = 0) {}   // error in C++20, OK in C++17
             ^
/Volumes/Store/Dropbox/Code/Projects/nimlab/lab/nimcache/@mcppinitissue.nim.cpp:41:10: note: in implicit initialization of field 'returnValue' with omitted initializer
FVector4 returnValue;
         ^
1 error generated.
Error: execution of an external compiler program 'clang++ -c -std=gnu++17 -funsigned-char  -pthread   -I/Volumes/Store/Nim/lib -I/Volumes/Store/Dropbox/Code/Projects/nimlab/lab -o /Volumes/Store/Dropbox/Code/Projects/nimlab/lab/nimcache/@mcppinitissue.nim.cpp.o /Volumes/Store/Dropbox/Code/Projects/nimlab/lab/nimcache/@mcppinitissue.nim.cpp' failed with exit code: 1
```

### Expected Output

```text
Ok
```

### Possible Solution

Dont emit `{}` 

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by heterodoxic**
fixes #21505

Amends the C++ codegen changes introduced by myself to ensure default constructors do not get nullified by Nim with #21378, **now** taking care of overload resolution rules regarding explicit constructors that were neglected in my first attempt: the generated C++ output should now be compilable again in the presence of explicit constructors.

### CODE REVIEW

**Comment by heterodoxic** [comment]
Will re-open this to force another CI run, if you do not mind: Two checks failed for reasons seemingly unrelated to my changes:

"(fatal: unable to access 'https://github.com/nim-lang/csources_v2.git/': Failed to connect to github.com port 443: Operation timed out)"

and:

"[Suite] SSL certificate check - sockets
Error: execution of an external program failed: '/Users/runner/.cache/nim/thttpclient_ssl_remotenetwork_d/thttpclient_ssl_remotenetwork_B17D345BCD37B5036DBA138E918D58131EB8B799'
/Users/runner/work/1/s/tests/misc/trunner_special.nim(22, 12): Check failed: ret == 0"

**Comment by ringabout** [comment]
Would you mind adding the test case from https://github.com/nim-lang/Nim/issues/21505? Btw, you might add `fixes #21505` in the title and body of your PR, which is documented in our contributing guide => https://nim-lang.github.io/Nim/contributing.html so that the linked issue will be closed automatically.

**Comment by heterodoxic** [comment]
Yes, absolutely! I will add a test case ASAP, possibly even a shorter one. Just need to figure out how to do this with this GitHub thingy, and read up on Nim's test conventions.
Sorry about the the title: I expected GitHub to parse it this way.

**Comment by Araq** [comment]
What's the hold up? Doesn't it improve the situation at least?

**Comment by heterodoxic** [comment]
Ah, sorry, I am just awfully taken hostage by some tricky RL stuff atm, and, unfortunately, my workaround dug up a general problem in how constructors may be codegen'ed especially for objects with explicit constructors and static storage duration in the context of aggregate initialization. It will result in differing behavior across the newer versions of GCC, Clang, and MSVC. Moreover, it depends on the respective C++ standard used: it might work for C++**14+** where some aggregate initialization rules were relaxed.

There is some stuff I need to reflect on first, as I do not want to crank out some ugly hacks like testing if `{{}}` was or is about to be emitted etc.

TL;DR: I am hopeful I will find the time and serenity to deliver a "proper solution" by this Sunday.

**Comment by Araq** [comment]
The perfect solution can wait but some fix to mitigate the regressions that your initial fix caused would be nice to have. :-)

**Comment by heterodoxic** [comment]
Really not happy with this, but it seems to help for now: I realize a much more rigorous effort is in order, so: WIP!
